cloud-controller-manager/do/metadata.go

```diff
@@ -17,27 +17,99 @@ limitations under the License.
 package do
 
 import (
+	"context"
 	"fmt"
 	"io/ioutil"
 	"net/http"
 	"os"
+	"time"
 
+	"github.com/digitalocean/godo"
+	"github.com/pkg/errors"
 	"k8s.io/klog/v2"
 )
 
-const dropletRegionMetadataURL = "http://169.254.169.254/metadata/v1/region"
+const (
+	dropletRegionMetadataURL = "http://169.254.169.254/metadata/v1/region"
+	resultsPerPage           = 50
+)
 
 // dropletRegion returns the region of the currently running program.
-func dropletRegion() (string, error) {
-	// FAKE_REGION can be specified for local testing.
-	if region := os.Getenv("FAKE_REGION"); region != "" {
-		klog.Warningf("Using fake region %q from environment variable FAKE_REGION -- this should only be set for testing purposes!", region)
-		return region, nil
+func dropletRegion(regionsService godo.RegionsService) (string, error) {
+	region := os.Getenv("REGION")
+	if region == "" {
+		return httpGet(dropletRegionMetadataURL)
+	}
+
+	validRegion, err := isValidRegion(region, regionsService)
+	if err != nil {
+		return "", fmt.Errorf("failed to determine if region is valid: %s", err)
+	}
+	if !validRegion {
+		return "", errors.New(fmt.Sprintf("invalid region specified: %s", region))
+	}
+
+	klog.Infof("Using region %q from environment variable", region)
+	return region, nil
+}
+
+// isValidRegion checks whether the given region is a valid DO region
+func isValidRegion(region string, regionsService godo.RegionsService) (bool, error) {
+	regions, err := listAllRegions(regionsService)
+	if err != nil {
+		return false, err
+	}
+
+	for _, reg := range regions {
+		if reg.Name == region {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func listAllRegions(regionsService godo.RegionsService) ([]godo.Region, error) {
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	list := make([]godo.Region, 0)
```

Comment:
I don't think the zero is needed. We can just do `make([]godo.Region)` or `var list []godo.Region`.